### DISCUSSION TITLE
Error when trying to make a signin page

### DISCUSSION THREAD

**Discussion by Ethan-Hill** [nextauthjs/next-auth#930] [Help] [answered]
Im having an issue with next-auth module
Im trying to create a custom signin page and this error keep happening 

```sh
Cannot convert undefined or null to object
    at Function.values (<anonymous>)
    at module.exports../pages/api/auth/signin.js.__webpack_exports__.default (C:\Users\Ethan\Documents\Development\Web Development\Full Stack\Next\discord-profile\.next\server\pages\api\auth\signin.js:121:22)
    at apiResolver (C:\Users\Ethan\Documents\Development\Web Development\Full Stack\Next\discord-profile\node_modules\next\dist\next-server\server\api-utils.js:8:7)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async DevServer.handleApiRequest (C:\Users\Ethan\Documents\Development\Web Development\Full Stack\Next\discord-profile\node_modules\next\dist\next-server\server\next-server.js:63:403)
    at async Object.fn (C:\Users\Ethan\Documents\Development\Web Development\Full Stack\Next\discord-profile\node_modules\next\dist\next-server\server\next-server.js:55:530)
    at async Router.execute (C:\Users\Ethan\Documents\Development\Web Development\Full Stack\Next\discord-profile\node_modules\next\dist\next-server\server\router.js:24:67)
    at async DevServer.run (C:\Users\Ethan\Documents\Development\Web Development\Full Stack\Next\discord-profile\node_modules\next\dist\next-server\server\next-server.js:65:1023)
    at async DevServer.handleRequest (C:\Users\Ethan\Documents\Development\Web Development\Full Stack\Next\discord-profile\node_modules\next\dist\next-server\server\next-server.js:33:450)
```
![Code_8ETDToqOhH](https://user-images.githubusercontent.com/45304268/101530290-9c44f600-3989-11eb-92e0-292611c8abed.png)
Any ideas why this might be happening?

`[...nextauth].js`
```js
import NextAuth from "next-auth";
import Providers from "next-auth/providers";

const options = {
  // Configure one or more authentication providers
  providers: [
    Providers.Discord({
      clientId: process.env.DISCORD_CLIENT_ID,
      clientSecret: process.env.DISCORD_CLIENT_SECRET,
    }),
  ],
  pages: {
    signIn: "/auth/signin",
    signOut: "/auth/signout",
    error: "/auth/error", // Error code passed in query string as ?error=
    verifyRequest: "/auth/verify-request", // (used for check email message)
    newUser: null, // If set, new users will be directed here on first sign in
  },
};

export default (req, res) => NextAuth(req, res, options);
```
`signin.js`
```js
import React from "react";
import { providers, signIn } from "next-auth/client";

export default ({ providers }) => {
  return (
    <>
      {Object.values(providers).map((provider) => (
        <div key={provider.name}>
          <button onClick={() => signIn(provider.id)}>
            Sign in with {provider.name}
          </button>
        </div>
      ))}
    </>
  );
};

export async function getInitalProps(context) {
  return {
    providers: await providers(context),
  };
}
```

**Comment by alaeddine119** [comment, marked as answer]
the `signin.js` is in `pages/api/auth/signin.js`, it should be located at `pages/auth/signin.js`. I hope this would fix it.

![Capture](https://user-images.githubusercontent.com/25179455/103603956-1b3d3800-4f10-11eb-967e-97f98be4f58c.PNG)

good luck !

**Reply by theindianappguy** [reply]
did this still not working showing error : https://next-auth.js.org/errors#client_fetch_error request to http://localhost:300/api/auth/providers failed, reason: connect ECONNREFUSED 127.0.0.1:300

**Reply by host2046** [reply]
you just need to close your project entirely and open the project and run it again

**Comment by prajwal-naik** [comment]
Did you find a solution to this problem?
My folder structure is similar to @alaeddine119 has suggested and my code is exactly the same as @Ethan-Hill 's but I'm still running into the same issue.
![image](https://user-images.githubusercontent.com/42693370/137676157-9b4ac775-7426-4130-a630-4b57342b860a.png)
![image](https://user-images.githubusercontent.com/42693370/137676208-740a622a-bf6a-4745-8594-e4daa9441e61.png)

**Reply by host2046** [reply]
you just need to close your project entirely and open the project and run it again

**Reply by host2046** [reply]
import { providers, signIn } from "next-auth/client";

const SignIn = ({ providers }) => {
  return (
    <>
      {providers &&
        Object.values(providers).map((provider) => (
          <div key={provider.name}>
            <button onClick={() => signIn(provider.id)}>
              Sign in with {provider.name}
            </button>
          </div>
        ))}
    </>
  );
};

SignIn.getInitialProps = async (context) => {
  return {
    providers: await providers(),
  };
};

export default SignIn;
 
this is the  updated version, you should add a conditional check to ensure that providers is not undefined before attempting to map over it. This should help prevent the "Cannot convert undefined or null to object" error.

**Comment by deadcoder0904** [comment]
I faced the same issue but in a different context. I was calling `getServerSideProps` outside of `pages/` ,i.e, in `components/` folder.

But I'd try calling `await getProviders()` [client-side](https://github.com/nextauthjs/next-auth/discussions/2978#discussioncomment-1489258) to see if it works or else a reproduction repo would help :)

**Reply by prajwal-naik** [reply]
I am trying to call it on the client-side itself, almost similar to the solution in the hyperlink in your answer :(

Here's the repo for reference, https://github.com/prajwal-naik/insta-clone.git

**Reply by deadcoder0904** [reply]
@prajwal-naik where is the client-side call? i don't see one. all i see is server-side call.

ps: don't expose your api keys like that. make sure to use environment variables. delete the git repo as the history remains.

**Reply by deadcoder0904** [reply]
@prajwal-naik it works for me. is this what you expected?

<img width="906" alt="Screenshot 2021-10-19 at 1 29 37 PM" src="https://user-images.githubusercontent.com/16436270/137868084-c0b31648-d364-485a-8ea8-aef66f814c10.png">

i got error on `next-auth/providers` so i changed the file `[...nextauth].js` to:

```js
import Google from 'next-auth/providers/google'

providers: [
		Google({
			clientId: process.env.GOOGLE_CLIENT_ID,
			clientSecret: process.env.GOOGLE_CLIENT_SECRET,
		}),
	],
```

**Reply by balazsorban44** [reply]
There is a mixed usage of v3 and v4 syntax, I recommend having a look at https://next-auth.js.org/getting-started/upgrade-v4

(I didn't see any exposed secrets though.)

**Reply by deadcoder0904** [reply]
https://github.com/prajwal-naik/insta-clone/blob/main/firebase.js has exposed secrets but idk as i dont use firebase so they might be public keys.

**Reply by balazsorban44** [reply]
They aren't conventional API keys. From docs (but I've also used Firebase myself), so OP didn't do anything wrong:

> Unlike how API keys are typically used, API keys for Firebase services are not used to control access to backend resources; that can only be done with Firebase Security Rules (to control which users can access resources) and App Check (to control which apps can access resources).

https://firebase.google.com/docs/projects/api-keys

**Reply by prajwal-naik** [reply]
> @prajwal-naik it works for me. is this what you expected?
> 
> <img alt="Screenshot 2021-10-19 at 1 29 37 PM" width="906" src="https://user-images.githubusercontent.com/16436270/137868084-c0b31648-d364-485a-8ea8-aef66f814c10.png">
> 
> i got error on `next-auth/providers` so i changed the file `[...nextauth].js` to:
> 
> ```js
> import Google from 'next-auth/providers/google'
> 
> providers: [
> 		Google({
> 			clientId: process.env.GOOGLE_CLIENT_ID,
> 			clientSecret: process.env.GOOGLE_CLIENT_SECRET,
> 		}),
> 	],
> ```

Yes, this is exactly what I wanted but for some reason, I am still getting the same issue. I even changed my [...nextauth].js file to reflect the code that you have specified.

![image](https://user-images.githubusercontent.com/42693370/137900228-b5157fa8-7565-42ed-a9df-5a04815cbcfb.png)
![image](https://user-images.githubusercontent.com/42693370/137900270-665cd599-2d9d-4b0a-963d-d23ab1f6edf5.png)

**Reply by prajwal-naik** [reply]
> They aren't conventional API keys. From docs (but I've also used Firebase myself), so OP didn't do anything wrong:
> 
> > Unlike how API keys are typically used, API keys for Firebase services are not used to control access to backend resources; that can only be done with Firebase Security Rules (to control which users can access resources) and App Check (to control which apps can access resources).
> 
> https://firebase.google.com/docs/projects/api-keys

I think @deadcoder0904 is talking about how exposing Firebase API keys can allow the public to spam a database?

**Reply by deadcoder0904** [reply]
@prajwal-naik maybe try clearing your `.next` cache & hard reload. other than that, idk probably!

**Reply by prajwal-naik** [reply]
No, didn't work then, either. 
Thanks, though.

**Reply by host2046** [reply]
this happens to me all the time
you just need to close your project entirely and open the project and run it again

**Comment by Abdul-Maajith** [comment]
Does Anyone knows the solution, As i encounter the same problem!
[Screenshot Capture - 2021-10-29 - 16-05-51](https://user-images.githubusercontent.com/77885682/139420723-23b8da84-3fc8-4081-9dfc-e668c258b71d.png)
m!

**Reply by theindianappguy** [reply]
i am stil facing this

**Comment by Ghastly2001** [comment]
> No, didn't work then, either. Thanks, though.

Don't worry I got you, buddy, you gotta replace your [...nextauth].js file content to 

import NextAuth from "next-auth"
import GoogleProvider from "next-auth/providers/google"

export default NextAuth({
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_SECRET,
    }),
    // ...add more providers here
  ],
  pages: {
    signIn: "/auth/signin"
  }

})

**Reply by harshdeep91** [reply]
bro u saved me

**Comment by Ghastly2001** [comment]
> i am stil facing this


Don't worry I got you, buddy, you gotta replace your [...nextauth].js file content to 

import NextAuth from "next-auth"
import GoogleProvider from "next-auth/providers/google"

export default NextAuth({
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_SECRET,
    }),
    // ...add more providers here
  ],
  pages: {
    signIn: "/auth/signin"
  }

})

**Comment by molikkidd** [comment]
I have the same issue when trying to create a custom signin page, When I console.log('list of providers', providers) inside of my getServerSideProps function, I can see the established providers in the console from my [...nextauth].js file but they are undefined in the SignIn() function. Seems to be a problem grabbing them from props.

**Comment by sayinmehmet47** [comment]
Will anybody solve this issue. I searched  a lot but couldn't find any solution 
![image](https://user-images.githubusercontent.com/75525090/147346614-70020f03-2ee3-495e-bc22-e79ac4368519.png)
![image](https://user-images.githubusercontent.com/75525090/147346635-c79c474f-38ad-4921-8129-ee122cba6887.png)

**Reply by AneeshaRama** [reply]
I have faced the same issue...But i solved it by adding new variable NEXTAUTH_SECRET=dbhjmrt^^bdbs*%^6t23(some random text).
Also make sure you have added NEXT_PUBLIC before your variable name like NEXT_PUBLIC_GOOGLE_CLIENT_ID
Hope it helps...

**Comment by ThangHuuVu** [comment]
Please check if you have set up the environment variable `NEXTAUTH_URL` correctly. Normally the value is `http://localhost:3000`:
```
NEXTAUTH_URL="http://localhost:3000/"
```

**Comment by Wanlleyy** [comment]
solução adicionar antes: providers &&, exemplo: 
{providers && Object.values(providers).map((provider) => (
    <div key={provider.name}>
    </div>
)};

**Reply by MarkusDavis** [reply]
Thank you so much!!!!!!! 

{providers && Object.values(providers).map((provider) => (
worked for me

**Comment by brunolucena** [comment]
Check if you're running your localhost in the same port as the one defined in your .env file
By default, it should probably be running on localhost:3000

`NEXTAUTH_URL=http://localhost:3000`

If you end up running the project on a different port, the providers will not be available.
